### ISSUE TITLE
[Proposal] `?=` operator: New If Nothing

### ISSUE THREAD

**Issue by VBAndCs** [dotnet/vblang#522]
I have a param `reqAttrs As List(Of String)` which I need to initialize it if nothing like this:

```VB.NET
reqAttrs = If(reqAttrs, New List(Of String))
```

I hope this can be shorten using null coalesce like this
```VB.NET
reqAttrs? = New List(Of String)
```

And I hope we don't mention the type and generic param since they are well known, I saw this proposal once, here or in C# repo and I hope it is done for all initializations:

```VB.NET
reqAttrs? = New
```

Or format it as:
```VB.NET
reqAttrs ?= New
```

Compare this to:
```VB.NET
reqAttrs = If(reqAttrs, New List(Of String))
```

If someone fears that `?=` is too symbolic and less readable, the alternative verbose VB one can be:

Or format it as:
```VB.NET
reqAttrs = New If Nothing
```

**Comment by rskar-git** [comment]
This sort of thing could be done in code:

    Function LazyNew(Of T As New)(<Out> ByRef v As T) As T
        If v Is Nothing Then v = New T
        Return v
    End Function
    Function LazyNew(Of T)(<Out> ByRef v As T, newOp As Func(Of T)) As T
        If v Is Nothing Then v = newOp()
        Return v
    End Function

...

        Dim reqAttrs As List(Of String) = Nothing
        Dim reqAttrs2 As List(Of String) = Nothing
        LazyNew(reqAttrs)
        LazyNew(reqAttrs2, Function() New List(Of String)(100))

I made them `Function` instead of `Sub` so that one could make them part of an expression, like so:

        Dim paramIdx = LazyNew(reqAttrs).IndexOf("ParamItem")

**Comment by VBAndCs** [comment]
Note that C# already allows this:
```C#
myList ??= new List<string>( );
```

I see no need to double `??`. One is enough in VB.

**Comment by Happypig375** [comment]
VB already has the two argument `If` operator, which is not made up of symbols. A `myList Nothing= New List(Of String)` will be better.

**Comment by VBAndCs** [comment]
VB also use the ? symbol to check nulls:
Dim c  = myList?.Count
So, we are familiar with the concept.

**Comment by DualBrain** [comment]
I'm not on board with this proposal... it just doesn't feel right when I look at it.  Even after having it explained to me, I still look at the suggested code and am still left wondering "Huh?".

reqAttrs = If(reqAttrs, New List(Of String))

This really isn't that much code and it leaves the future reader with little to no confusion as to what is taking place.

**Comment by VBAndCs** [comment]
@DualBrain
In my opinion, the `If(x, y)` is not a VB style syntax, and you can't understand what it does unless you learn it first. Your argument dictates that we should use a full `If Else` block for readability purpose, or at least:
`If reqAttrs Is Nothing Then reqAttrs = New List(Of String))`
This is the least you can offer for future readers.
But, your purpose can be achieved easily even with:
`reqAttrs ?= new`
just by hovering over `?=` to view a tooltip showing: 
> If Nothing then initialize with ... 

and by hovering over `New` to see a tip showing: `New List(Of String))`
It is a must to consider Intellisense as a part of the language, to get rid of any unnecessary verbose syntax. I wrote code before Intellisense appeared in VB5, and I feel the grace of having it. VB syntax after, is not necessarily to be as before. I think we are 32 years late to use the Intellisense revolution in VB.NET!

**Comment by rskar-git** [comment]
>In my opinion, the `If(x, y)` is not a VB style syntax,...

Yeah, I agree, it isn't obvious. But in it's defense, just by hovering over it, there is a tooltip that tells you. C# has the `??` as the equivalent, and because it's a symbol and not a keyword, `??=` is a natural part of its collection of operators. Anyway, a "lazy-loader" based on the C# `??=` concept is somewhat interesting, but can be emulated easily enough (such as the `LazyNew` functions suggested above).

**Comment by DualBrain** [comment]
There is one big difference between ?= and If(x, y).  If, to a BASIC dev is a keyword/function.  Just like *ALL* keywords/functions/methods - yes, the reader has to be knowledgeable about what they ultimately do - but, you can easily read what it does.  If( x {or} y) -> hmmm... why would it pick one or the other -> read manual.  

reqAttrs? = New List(Of String))

This syntax doesn't tell me that it will only set the value on the left to the value on the right if the value on the left is currently nothing.  It tells me that I have a value on the left that can be nullable being set to a new value on the right given the existing rules that are in place.  (Yes, this is a compiler error, but my point is that when I see it written this way this is what I'm thinking.)

reqAttrs ?= New List(Of String))

This is better, but... as you point out... but, it is syntactic sugar for...

reqAttrs = If(reqAttrs, New List(Of String))

Already works... already available... doesn't require learning anything new and, ultimately, isn't that much to type.  Then you suggest...

reqAttrs = New If Nothing 

Which I would assume would really be (see below)...

reqAttrs = New List(Of String)) If Nothing

Which really isn't that much of a savings over the existing available method of accomplish the same task.  Without the List(Of String)) portion, the code looks like it's just a few keywords randomly placed on a line to create an expression.  If it does require the full line as I present... what is gained?

Regarding If(x, y) and, by extension If(b, x, y), are very established VB approaches and... have a long history when you consider IIF.  And, as for them not being "VB"... they are every bit as VB as MID.

The biggest problem with this proposal is, as far as I see, is that there really isn't a lot to be gained by changing the language for this scenario.  We already have a way to initialize an object if it is currently empty using a relatively short syntax that is pretty clear in it's intent and readability.  The most attractive of the suggestions in this thread is a savings of 12 keystrokes... doesn't reduce the line count and, ultimately, I don't believe significantly improves the readability over the existing method.  However, I could get behind it for the simple "12 keystrokes" savings... I just don't know that it's the hill that I'd be willing to die on given that the alternative isn't really that much more effort. ;-)
